### PR TITLE
forecasting: infer cross-channel mode from checkpoint

### LINKED ISSUE #63: perform_forecasting silently runs randomly initialised cross-channel weights with custom checkpoints
When a custom `ckpt` path is passed to `perform_forecasting()`, the SDK builds the model **with** the cross-channel attention block but loads a checkpoint that has **no** weights for it. Those parameters keep their random initialisation and participate in every forward pass. No exception, no warning.

The most compact symptom: **the same checkpoint file produces different forecasts depending on whether you pass it by name or by absolute path.**

## Environment

- NV-Tesseract @ `57658bd760846497f02f8c5c15decc2be38e8e92`
- torch 2.13.0, Python 3.12
- Reproduced on both CUDA 13.2 / aarch64 and Apple MPS

## Reproduction

Only the published weights are needed — no fine-tuning. Run from `forecasting/`:

```python
import os

import numpy as np
import pandas as pd

from sdk.forecasting import perform_forecasting

rng = np.random.default_rng(0)
df = pd.DataFrame({
    "timestamp": pd.date_range("2000-01-01", periods=512, freq="D"),
    "target": np.log(rng.lognormal(3.0, 0.3, 512)),
})

def run(**kw):
    out = perform_forecasting(df=df, seq_len=512, forecast_horizon=5, **kw)
    return np.round(out["target_forecast"].to_numpy(), 4)

by_name = run(ckpt="moment_head_512_6hr.pt", use_cross_channel=False)
path = os.path.abspath("moment_head_512_6hr.pt")   # the very same file
by_path = run(ckpt=path)
by_path_off = run(ckpt=path, use_cross_channel=False)

print("by name            :", by_name)
print("by absolute path   :", by_path)
print("by path, cc=False  :", by_path_off)
```

Observed:

```
by name            : [3.0587 3.157  3.3334 3.2909 3.1907]
by absolute path   : [3.302  4.2455 5.4895 5.3521 4.2696]
by path, cc=False  : [3.0587 3.157  3.3334 3.2909 3.1907]
```

## Root cause

Three individually reasonable decisions compose into a silent failure.

**1. The SDK layer inverts the model layer's default.** `sdk/forecasting.py:2157` declares `perform_forecasting(use_cross_channel: bool = True)`, and `sdk/forecasting.py:118` does the same for `_load_cached_model`. `build_model` in `model.py:28` defaults to `False`.

**2. The compensating checkpoint swap is a string comparison.** `sdk/forecasting.py:2291-2292`:

```python
if ckpt == DEFAULT_CHECKPOINT_NAME and use_cross_channel:
    ckpt = DEFAULT_CROSS_CHANNEL_CHECKPOINT_NAME
```

With the default name this substitutes `run8_best_model_cr.pt`, which does carry cross-channel weights, and everything is consistent. With any other `ckpt` value — including an absolute path to that same default file — the branch does not fire, yet `use_cross_channel` remains `True`, so the block is still built.

**3. Missing cross-channel keys are explicitly exempted.** `sdk/forecasting.py:153-161`:

```python
load_result = model.load_state_dict(state, strict=False)
missing = list(getattr(load_result, "missing_keys", [])) if load_result is not None else []
unexpected = list(getattr(load_result, "unexpected_keys", [])) if load_result is not None else []
if unexpected:
    raise RuntimeError(f"Unexpected checkpoint keys for forecasting model: {unexpected}")
if missing:
    non_cr_missing = [key for key in missing if "cross_channel" not in key]
    if non_cr_missing:
        raise RuntimeError(f"Missing non-cross-channel checkpoint keys: {non_cr_missing}")
```

Unexpected keys raise and missing keys raise — except `cross_channel*`, which is filtered out. The block therefore runs with its random initialisation.

## Impact

Any call passing a custom `ckpt` without explicitly setting `use_cross_channel=False`. That covers **inference with fine-tuned checkpoints in general**: `examples/finetune_example.py` does not enable `--use-cross-channel` by default, so its output never contains these weights, and neither the SDK README nor the example notes that the flag has to be turned off at inference.

This is easy to mistake for a modelling result rather than a bug:

- Fine-tuning is unaffected — it does not go through this load path — so training and validation curves look healthy.
- `control_randomness(seed=seed)` makes the random initialisation reproducible, so the wrong output is stable across runs and reads as genuine model behaviour.
- The random block still consumes its input, so outputs correlate with the input rather than looking like noise.

In my own use, a head fine-tuned on a univariate financial series reached a healthy validation MSE, predicted in-sample correctly, and produced out-of-sample errors 17–26x the naive baseline. That combination is the textbook signature of overfitting, and I nearly recorded it as one. What gave it away was the magnitude: the model emitted values far outside the range present anywhere in its training data, which overfitting does not explain.

Possibly related but distinct: #31 reports fine-tuned AD checkpoints not replaying training-time preprocessing — a different mechanism in a different module, but the same general shape of fine-tuning artifacts not being honoured by the SDK inference path.

## Suggested fix

Resolve the flag from the checkpoint contents rather than from the caller: inspect the state dict for `cross_channel*` keys before `build_model`, warn on disagreement, and follow the checkpoint. That makes "block present, weights random" structurally unreachable instead of relying on callers passing the right argument, and the missing-key exemption in step 3 can then be dropped, since the block is only built when its weights are present.

Lighter alternatives, for comparison: emit a warning when cross-channel keys are actually filtered out; or document that custom checkpoints require `use_cross_channel=False`. Both leave the invalid state reachable.

**PR in the works.** I have the checkpoint-driven version implemented and verified locally: the reproduction above returns identical values on all three paths, and `sdk/tests/` passes (60 passed). It also updates `test_perform_forecasting_uses_cross_channel_model_by_default`, which asserts the current contract — the mocked `torch.load` returns `{}`, so under the fix that checkpoint correctly resolves to no cross-channel block. The updated test preserves the original intent by giving the mocked checkpoint cross-channel weights, and two new tests cover both directions of the resolution. Happy to open it if the approach looks right.

### PR DESCRIPTION
**Pull request by tzhengus**
## Summary

- inspect checkpoint weights before constructing the forecasting model
- derive cross-channel mode from `cross_channel_attn.*` keys and warn when the requested mode disagrees
- reject all missing checkpoint keys and reuse the same cached model across caller mode hints
- add regression coverage for standard, cross-channel, incomplete, and cached checkpoint loading

## Testing

- `HF_HUB_OFFLINE=1 TRANSFORMERS_OFFLINE=1 .venv/bin/python -m pytest sdk/tests examples/tests -q --tb=short` (66 passed)
- `forecasting/.venv/bin/ruff check forecasting ad_diffusion --config forecasting/pyproject.toml`
- `forecasting/.venv/bin/ruff format --check forecasting ad_diffusion --config forecasting/pyproject.toml`
- `python3 scripts/add_spdx_headers.py --check`

Fixes #63

### CODE REVIEW

**Comment by neptunel** [comment]
Hi @tzhengus , thanks for your contribution. Your suggestion with few other alterations are merged in #73 .
